specs/experimental/custom-gas-token.md

```diff
@@ -0,0 +1,391 @@
+
+Both the L1 and L2 smart contract systems MUST be able to introspect on the address of the gas paying token to
+ensure the security of the system. The source of truth for the configuration of the address of the token is
+in the L1 `SystemConfig` smart contract.
+
+The terms "custom gas token", "gas paying token", "native asset" and "gas paying asset" can all be used interchangebly.
+Ether is not a token.
+
+## Constants
+
+| Constant          | Value                           | Description |
+| ----------------- | ------------------------------- | ----------- |
+| `ETHER_TOKEN_ADDRESS` |  `address(0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE)` | Represents ether for gas paying asset |
+| `DEPOSITOR_ACCOUNT` | `address(0xDeaDDEaDDeAdDeAdDEAdDEaddeAddEAdDEAd0001)` | Account with auth permissions on `L1Block` contract |
+| `GAS_PAYING_TOKEN_SLOT` | `bytes32(uint256(keccak256("opstack.gaspayingtoken")) - 1)` | Storage slot that contains the address and decimals of the gas paying token |
+
+## Properties of a Gas Paying Token
+
+The gas paying token MUST satisfy the standard [ERC20](https://eips.ethereum.org/EIPS/eip-20) interface
+and implementation. It MUST be a L1 contract.
+
+The gas paying token cannot:
+
+- Have a fee on transfer
+- Have rebasing logic
+- Have call hooks on transfer
+- Have more than 18 decimals
+- Have out of band methods for modifying balance or allowance
+
+It MUST be enforced on chain that the token has 18 or less decimals to guarantee no losses of precision when
+depositing a token that has a different number of decimals.
+
+### Automated Validation
+
+Ideally it is possible to have automated validation of custom gas paying tokens to know that the token satisfies the
+above properties. Due to the nature of the EVM, it may not always be possible to do so. [ERC-165][erc165]
+isn't always used by ERC20 tokens and isn't guaranteed to tell the truth. [USDT](usdt)
+does not correctly implement the ERC20 spec. It may be possible to use execution traces to observe the
+properties of the ERC20 tokens but some degree of social consensus will be required for determining
+the validity of an ERC20 token.
+
+[erc165]: https://eips.ethereum.org/EIPS/eip-165
+[usdt]: https://forum.openzeppelin.com/t/usdt-locked-in-a-contract-that-has-a-withdraw-token-function/32154/2
+
+## Configuring the Gas Paying Token
+
+The gas paying token is set within the L1 `SystemConfig` smart contract. This allows for easy access to the information
+required to know if an OP Stack chain is configured to use a custom gas paying token. The gas paying token is set
+during initialization and cannot be modified by the `SystemConfig` bytecode. Since the `SystemConfig` is proxied,
+it is always possible to modify the storage slot that holds the gas paying token address directly during an upgrade.
+It is assumed that the chain operator will not modify the gas paying token address unless they specifically decide to
```

Comment:
how does the system cope with a change of the gas paying token? would anything break? especially curious how users would experience this change. additionally, it would be interesting to explore how an already-deployed chain handles going from eth to custom token

Reply:
We don't necessarily care about that as a feature when we go to market so while it may be an interesting thing, we can say its out of scope. If a team wants to contribute a how to guide on this then we could accept it in the future